File: app-web/vela-wallet/src/lib/wallet/core/tracker-executor.ts
// Ported from src/services/wallet-state-core/tx-tracker-executor.ts @ f9bcb278 — RN seams rewritten to the web modules; logic verbatim.
/**
 * The only place the `tx_tracker` core touches the outside world.
 *
 * Six operations, each one existing service call. No branching on business
 * meaning: every throttle (3 s receipt, 12 s status, 12 s single-flight
 * reconcile), the 120 s window, the 24 h abandon line and every verdict live in
 * Rust. What lives HERE is exactly what the core's module doc assigns to the
 * shell:
 *
 * - **The whole wording/instanceof layer.** `requestUserOpReceipt` already
 *   collapses `rpcCall` into the typed axis the core wants — an RPC error or a
 *   throw is `reachedBundler: false` (→ `ReceiptUnreachable`, NEVER a failure),
 *   no result or no `transactionHash` is `resolution: null` (→ `ReceiptPending`),
 *   `success !== false` is `confirmed` (→ `Receipt`) and `success === false` is
 *   `failed` (→ `ReceiptFailed`). `pollUserOpStatus` answers `null` for a null
 *   result, an RPC error or an older relay that has no such method (→
 *   `StatusUnavailable`). So the regexes that used to *be* the classification
 *   (`/dropped from the network/`, `instanceof UserOpRejectedError`) have no
 *   remaining reader on web.
 * - **The clock.** Every time-bearing result carries `now_ms`; the core owns
 *   cadence but never reads a clock.
 * - **The shared 3 s receipt throttle (invariant ⑤).** Going through
 *   `requestUserOpReceipt` rather than a raw `rpcCall` is load-bearing: the
 *   receipt sheet (`TransactionReceipt.tsx`), the detail sheet and
 *   `safe-transaction.ts`'s own `waitForReceipt` join the SAME in-flight request
 *   and the SAME cooldown, so a hash watched by four surfaces still costs one
 *   `eth_getUserOperationReceipt` every 3 s. The core coalesces its own side
 *   (one in-flight request per hash); this cache coalesces across machines.
 * - **The receipt's own by-products.** `NotifyConfirmed` carries no payload
 *   beyond the hash — the AUTHENTIC logs and the sender are facts the shell
 *   already holds from the poll it just made, and they are what `token_trust`
 *   admits tokens from. They are cached per hash here and consumed once.
 *
 * Failure contract (shared effect loop): nothing rejects. Every rejection is
 * converted into the result variant that operation answers with — and for this
 * machine the "failure" variants are all the honest-unknown ones, never a
 * verdict.
 */

import { loadTransactions, updateTransactions } from '$lib/services/records';
import type { LocalTransaction } from '$lib/services/transactions-model';
import { pollUserOpStatus, requestUserOpReceipt } from '$lib/services/tx-reconciler';

import type { TrackLifecycle } from '$lib/core/generated/TrackLifecycle';
import type { TrackPendingRecord } from '$lib/core/generated/TrackPendingRecord';
import type { TrackShellResult } from '$lib/core/generated/TrackShellResult';
import type { TrustReceiptLog } from '$lib/core/generated/TrustReceiptLog';
import type { TrackEffect, TrackShellPorts } from './tracker-types';

/**
 * The two record kinds a UserOp submission produces — `send` (the send screen,
 * one row per batch recipient) and `dapp_tx` (the signing sheet). `type` is
 * optional on older rows and defaults to `send`, exactly as `storage.ts` says.
 * Everything else in the store (receives, signatures, connections) has no
 * userOpHash to converge.
 */
const TRACKED_TYPES = new Set(['send', 'dapp_tx']);

/**
 * Per-hash by-products of the last receipt poll. Bounded because a long-lived
 * tab can watch many ops; the cap is the same order as the sign resident's
 * transport table, and the entries are consumed by their own `NotifyConfirmed`.
 */
const MAX_TRACKED_RECEIPTS = 64;

/** `LocalTransaction.status`, narrowed to what a patch may write. */
type PatchStatus = 'confirmed' | 'failed';

function normalize(hash: string): string {
	return hash.toLowerCase();
}

/** A receipt log, kept only if it has the three fields the netting reads. */
function toTrustLog(log: unknown): TrustReceiptLog | null {
	const candidate = log as { address?: unknown; topics?: unknown; data?: unknown } | null;
	if (!candidate || typeof candidate.address !== 'string') return null;
	if (!Array.isArray(candidate.topics)) return null;
	return {
		address: candidate.address,
		topics: candidate.topics.filter((topic): topic is string => typeof topic === 'string'),
		data: typeof candidate.data === 'string' ? candidate.data : '0x'
	};
}

/**
 * The still-pending submissions the reconcile sweep answers with — the union of
 * the two scans it replaces: `tx-reconciler.ts:217-224` (any `send`) and the
 * dApp startup scan (`dapp-connection.tsx:1038-1040`, `dapp_tx`).
 *
 * Deliberately NOT filtered by account: the TS reconciler filtered on the Home
 * account because it was called with one, and a pending op belonging to another
 * local account was simply never converged. The core keys everything by hash and
 * patches records by id, so the honest superset is the right answer — see the
 * behaviour note in the integration report.
 *
 * The 24 h line is left to the core (invariant ④): it drops records that old
 * itself, so the filter here stays a pure "is this a live submission" read.
 */
function toPendingRecords(txs: LocalTransaction[]): TrackPendingRecord[] {
	const records: TrackPendingRecord[] = [];
	for (const tx of txs) {
		if (tx.status !== 'pending') continue;
		if (!tx.userOpHash) continue;
		if (tx.txHash !== '') continue; // already confirmed on-chain
		if (!TRACKED_TYPES.has(tx.type ?? 'send')) continue;
		records.push({
			record_id: tx.id,
			user_op_hash: tx.userOpHash,
			chain_id: tx.chainId,
			// Stored in SECONDS; the core measures every deadline in epoch ms.
			submitted_at_ms: tx.timestamp * 1000
		});
	}
	return records;
}

export function createTxTrackerExecutor(ports: TrackShellPorts) {
	/** hash → the authentic logs of the receipt that confirmed it. */
	const logsByHash = new Map<string, TrustReceiptLog[]>();
	/** hash → the account that submitted it (`LocalTransaction.from`). */
	const fromByHash = new Map<string, string>();

	function remember<T>(map: Map<string, T>, key: string, value: T): void {
		if (map.size >= MAX_TRACKED_RECEIPTS && !map.has(key)) {
			const oldest = map.keys().next().value;
			if (oldest !== undefined) map.delete(oldest);
		}
		map.set(key, value);
	}

	/**
	 * Who received what this op delivered. The submitting account is
	 * `LocalTransaction.from` — the very value `autoAddReceivedTokens(tx.from, …)`
	 * and `dapp-connection.tsx:786-802` passed. Cached from the reconcile sweep;
	 * otherwise read back from the store, which is where the record was written
	 * before this op could ever be tracked (send: `RecordsPersisted` precedes
	 * `TrackSubmitted`; dApp: §4's record precedes the receipt wait).
	 *
	 * No sender, no auto-add: token_trust is never handed a guess.
	 */
	async function senderOf(hash: string): Promise<string | null> {
		const cached = fromByHash.get(hash);
		if (cached) return cached;
		const txs = await loadTransactions().catch(() => []);
		const match = txs.find((tx) => !!tx.userOpHash && normalize(tx.userOpHash) === hash);
		if (!match) return null;
		remember(fromByHash, hash, match.from);
		return match.from;
	}

	async function execute(effect: TrackEffect): Promise<TrackShellResult> {
		const operation = effect.operation;
		switch (operation.type) {
			case 'now':
				return { type: 'clock', now_ms: Date.now() };

			case 'poll_receipt': {
				const hash = normalize(operation.user_op_hash);
				// Never throws: its own try/catch already answers `reachedBundler:false`.
				const outcome = await requestUserOpReceipt(operation.user_op_hash, operation.chain_id);
				const now_ms = Date.now();
				if (!outcome.reachedBundler) {
					// Timeout / network / RPC error. NOT a failure (invariant ①), and
					// honestly distinct from "the bundler answered, nothing yet" (⑧).
					return { type: 'receipt_unreachable', user_op_hash: hash, now_ms };
				}
				const resolution = outcome.resolution;
				if (!resolution) return { type: 'receipt_pending', user_op_hash: hash, now_ms };
				const txHash = resolution.txHash ?? '';
				if (!txHash) return { type: 'receipt_pending', user_op_hash: hash, now_ms };
				if (resolution.failed) {
					// `success === false` — dropped or reverted. The one receipt shape
					// that may fail a record (invariant ③).
					return { type: 'receipt_failed', user_op_hash: hash, tx_hash: txHash, now_ms };
				}
				// Hold the authentic logs for the `NotifyConfirmed` that follows the
				// patch — they are the only thing token_trust may auto-add from.
				const logs = (resolution.logs ?? [])
					.map(toTrustLog)
					.filter((log): log is TrustReceiptLog => log !== null);
				remember(logsByHash, hash, logs);
				// With the authentic logs: the core decides whether the Safe inside
				// the op actually executed (spec 038 #D1 — `ExecutionFailure` under
				// a `success: true` is a failed payment, not a confirmed one).
				return { type: 'receipt_with_logs', user_op_hash: hash, tx_hash: txHash, now_ms, logs };
			}

			case 'poll_status': {
				const hash = normalize(operation.user_op_hash);
				// Never throws either: a null/error/older-relay answer is `null`.
				const status = await pollUserOpStatus(operation.user_op_hash, operation.chain_id);
				const now_ms = Date.now();
				if (!status) return { type: 'status_unavailable', user_op_hash: hash, now_ms };
				return {
					type: 'status',
					user_op_hash: hash,
					status: status.status as TrackLifecycle,
					stage: status.stage ?? null,
					now_ms
				};
			}

			case 'load_pending_txs': {
				// `loadTransactions().catch(() => [])`, verbatim: a store that cannot be
				// read answers an empty sweep, never a fault.
				const txs = await loadTransactions().catch(() => []);
				const records = toPendingRecords(txs);
				for (const tx of txs) {
					if (tx.userOpHash) remember(fromByHash, normalize(tx.userOpHash), tx.from);
				}
				return { type: 'records_loaded', records, now_ms: Date.now() };
			}

			case 'update_tx_records': {
				const status: PatchStatus = operation.patch.status === 'failed' ? 'failed' : 'confirmed';
				const patch: Partial<LocalTransaction> =
					operation.patch.tx_hash != null
						? { status, txHash: operation.patch.tx_hash }
						: { status };
				// ONE atomic read-modify-write for every sibling of a batch, same ids,
				// in place — never a second record (invariant ⑦). Best effort, as every
				// TS call site's `.catch(() => {})` was.
				await updateTransactions(operation.ids, patch).catch(() => {});
				// `activity_feed`'s `ReconcileCompleted`: verdicts landed, so the feed
				// re-reads the store — and deliberately does not celebrate them.
				ports.feedReconciled(operation.ids.length);
				return { type: 'records_patched' };
			}

			case 'notify_confirmed': {
				const hash = normalize(operation.user_op_hash);
				const logs = logsByHash.get(hash);
				logsByHash.delete(hash);
				if (logs && logs.length > 0) {
					const from = await senderOf(hash);
					if (from) ports.receiptLogsConfirmed(from, operation.chain_id, logs);
				}
				fromByHash.delete(hash);
				return { type: 'notified' };
			}
		}
	}

	/**
	 * The defensive tail. Every operation above swallows its own errors, so this
	 * is only ever reached by a fault in the plumbing — and for this machine the
	 * only safe answers are the ones that keep money in flight *pending*.
	 */
	function toFailure(effect: TrackEffect): TrackShellResult {
		const operation = effect.operation;
		const now_ms = Date.now();
		switch (operation.type) {
			case 'now':
				return { type: 'clock', now_ms };
			case 'poll_receipt':
				// Unreachable, never failed (invariant ①).
				return {
					type: 'receipt_unreachable',
					user_op_hash: normalize(operation.user_op_hash),
					now_ms
				};
			case 'poll_status':
				return {
					type: 'status_unavailable',
					user_op_hash: normalize(operation.user_op_hash),
					now_ms
				};
			case 'load_pending_txs':
				return { type: 'records_loaded', records: [], now_ms };
			case 'update_tx_records':
				// The core must still be told, or the entry never leaves its
				// patch-pending state — and the feed is told nothing changed.
				return { type: 'records_patched' };
			case 'notify_confirmed':
				return { type: 'notified' };
		}
	}

	return { execute, toFailure };
}

File: app-ios/VelaWallet/VelaWallet/Features/Send/TrackerExecutor.swift
//
//  TrackerExecutor.swift
//  VelaWallet
//
//  The only place the `tx_tracker` core touches the outside world.
//
//  Six operations, ported from `app-android/.../feature/send/core/
//  TrackerExecutor.kt` (spec 043). Every cadence decision is the core's — how
//  often to poll, how long to wait, when to give up — and this supplies the
//  clock, the transport and the notification.
//
//  ## The receipt's logs are the ONLY way a token is ever admitted
//
//  `notify_confirmed` hands the logs this executor just polled to
//  `token_trust::ReceiptLogsConfirmed`. That is the founder's standing ruling:
//  a token is admitted from a **confirm-time receipt**, never from a sign-time
//  simulation, because a simulation is something a site can author and a
//  receipt is not.
//

import Foundation

@MainActor
final class TrackerExecutor {

    /// Every operation this executor is required to handle.
    static let operations = [
        "poll_receipt", "poll_status", "load_pending_txs",
        "update_tx_records", "notify_confirmed", "now",
    ]

    /// What the app owns: the notification, and the auto-add hand-off.
    struct Ports {
        var notifyConfirmed: (_ userOpHash: String, _ chainId: Int, _ txHash: String) -> Void = { _, _, _ in }
        /// The authentic logs, on their way to `token_trust`.
        var receiptLogs: (_ from: String, _ chainId: Int, _ logs: [[String: Any]]) -> Void = { _, _, _ in }
        /// Records changed on disk — the feed re-reads.
        var recordsPatched: () -> Void = {}
    }

    private let store: VelaStore
    private let relay: RelayClient
    var ports: Ports

    /// The logs of a receipt already polled, kept until `notify_confirmed`
    /// asks for them. The core routes the two as separate operations, and
    /// re-polling would be a second round trip for an answer already in hand.
    private var logsByHash: [String: (sender: String, logs: [[String: Any]])] = [:]

    init(store: VelaStore, relay: RelayClient, ports: Ports = Ports()) {
        self.store = store
        self.relay = relay
        self.ports = ports
    }

    func perform(_ operation: [String: Any]) async -> String {
        let hash = operation["user_op_hash"] as? String ?? ""
        let chainId = (operation["chain_id"] as? NSNumber)?.intValue ?? 0

        switch operation["type"] as? String ?? "" {

        case "poll_receipt":
            return await pollReceipt(hash: hash, chainId: chainId)

        case "poll_status":
            guard let answer = await relay.userOpStatus(chainId: chainId, userOpHash: hash) else {
                // An older relay, or one that could not be asked. Not a verdict.
                return CoreJSON.string([
                    "type": "status_unavailable", "user_op_hash": hash, "now_ms": Self.nowMs,
                ])
            }
            return CoreJSON.string([
                "type": "status",
                "user_op_hash": hash,
                "status": answer.status,
                "stage": answer.stage.map { $0 as Any } ?? NSNull(),
                "now_ms": Self.nowMs,
            ])

        case "load_pending_txs":
            // Derived from the store, which is why a force-quit loses nothing.
            let records = TxRecords.pending(store: store).compactMap(Self.pendingWire)
            return CoreJSON.string([
                "type": "records_loaded", "records": records, "now_ms": Self.nowMs,
            ])

        case "update_tx_records":
            let ids = operation["ids"] as? [String] ?? []
            let patch = operation["patch"] as? [String: Any] ?? [:]
            var fields: [String: Any] = ["status": patch["status"] as? String ?? "confirmed"]
            if let txHash = patch["tx_hash"] as? String, !txHash.isEmpty {
                fields["txHash"] = txHash
            }
            TxRecords.patch(ids: ids, fields: fields, store: store)
            ports.recordsPatched()
            return CoreJSON.string(["type": "records_patched"])

        case "notify_confirmed":
            let txHash = operation["tx_hash"] as? String ?? ""
            ports.notifyConfirmed(hash, chainId, txHash)
            // The authentic logs, to the one entry point that may admit a
            // token. Nothing else in the app is allowed to.
            if let held = logsByHash.removeValue(forKey: hash), !held.logs.isEmpty {
                ports.receiptLogs(held.sender, chainId, held.logs)
            }
            return CoreJSON.string(["type": "notified"])

        case "now":
            return CoreJSON.string(["type": "clock", "now_ms": Self.nowMs])

        default:
            print("[vela-wallet] tx_tracker: unhandled operation \(operation["type"] ?? "?")")
            return Self.neutralAnswer(operation)
        }
    }

    /// The core's own mapping, spelled out in its doc and not re-decided here:
    /// an RPC error or a throw is **unreachable**; no result or no transaction
    /// hash is **pending**; `success != false` is a receipt; `success == false`
    /// is a failure.
    private func pollReceipt(hash: String, chainId: Int) async -> String {
        switch await relay.userOpReceipt(chainId: chainId, userOpHash: hash) {
        case .unreachable:
            return CoreJSON.string([
                "type": "receipt_unreachable", "user_op_hash": hash, "now_ms": Self.nowMs,
            ])
        case .pending:
            return CoreJSON.string([
                "type": "receipt_pending", "user_op_hash": hash, "now_ms": Self.nowMs,
            ])
        case .resolved(let confirmed, let txHash, let sender, let logs):
            guard confirmed else {
                return CoreJSON.string([
                    "type": "receipt_failed", "user_op_hash": hash,
                    "tx_hash": txHash, "now_ms": Self.nowMs,
                ])
            }
            logsByHash[hash] = (sender ?? "", logs)
            // `receipt_with_logs` rather than `receipt`: the core reads the
            // logs for an `ExecutionFailure` inside an operation the
            // EntryPoint counted as a success — the money did not move and the
            // record must say `failed`, not `confirmed`.
            return CoreJSON.string([
                "type": "receipt_with_logs", "user_op_hash": hash,
                "tx_hash": txHash, "now_ms": Self.nowMs, "logs": logs,
            ])
        }
    }

    /// One stored row as `TrackPendingRecord`.
    static func pendingWire(_ record: [String: Any]) -> [String: Any]? {
        guard let id = record["id"] as? String, !id.isEmpty,
              let hash = record["userOpHash"] as? String, !hash.isEmpty
        else { return nil }
        return [
            "record_id": id,
            "user_op_hash": hash,
            "chain_id": (record["chainId"] as? NSNumber)?.intValue ?? 0,
            // Stored in SECONDS; the core counts in milliseconds.
            "submitted_at_ms": ((record["timestamp"] as? NSNumber)?.doubleValue ?? 0) * 1000,
        ]
    }

    static func neutralAnswer(_ operation: [String: Any]) -> String {
        let hash = operation["user_op_hash"] as? String ?? ""
        switch operation["type"] as? String ?? "" {
        case "poll_receipt":
            return CoreJSON.string([
                "type": "receipt_unreachable", "user_op_hash": hash, "now_ms": nowMs,
            ])
        case "poll_status":
            return CoreJSON.string([
                "type": "status_unavailable", "user_op_hash": hash, "now_ms": nowMs,
            ])
        case "load_pending_txs":
            return CoreJSON.string(["type": "records_loaded", "records": [], "now_ms": nowMs])
        case "update_tx_records":
            return CoreJSON.string(["type": "records_patched"])
        case "notify_confirmed":
            return CoreJSON.string(["type": "notified"])
        default:
            return CoreJSON.string(["type": "clock", "now_ms": nowMs])
        }
    }

    private static var nowMs: Double { Date().timeIntervalSince1970 * 1000 }
}
